src/tools/debugger/session/Breakpoints.java

```diff
@@ -57,52 +69,62 @@ public Breakpoint addOrUpdate(final LineBreakpoint bId) {
     bp.setEnabled(bId.isEnabled());
-    return bp;
   }
 
-  public Breakpoint addOrUpdate(final MessageSenderBreakpoint bId) {
-    Breakpoint bp = truffleBreakpoints.get(bId);
-    if (bp == null) {
-      WebDebugger.log("SetSectionBreakpoint: " + bId);
-      bp = Breakpoint.newBuilder(bId.getCoordinate().uri).
-          lineIs(bId.getCoordinate().startLine).
-          columnIs(bId.getCoordinate().startColumn).
-          sectionLength(bId.getCoordinate().charLength).
-          build();
-      debuggerSession.install(bp);
-      truffleBreakpoints.put(bId, bp);
+  public synchronized void addOrUpdate(final MessageSenderBreakpoint bId) {
+    saveTruffleBasedBreakpoints(bId);
+  }
+
+  public synchronized void addOrUpdate(final AsyncMessageReceiverBreakpoint bId) {
+    saveTruffleBasedBreakpoints(bId);
+  }
+
+  public synchronized void addOrUpdate(final MessageReceiverBreakpoint bId) {
+    FullSourceCoordinate coord = bId.getCoordinate();
+    BreakpointEnabling<MessageReceiverBreakpoint> existingBP = receiverBreakpoints.get(coord);
+    if (existingBP == null) {
+      existingBP = new BreakpointEnabling<MessageReceiverBreakpoint>(bId);
+      receiverBreakpoints.put(coord, existingBP);
+    } else {
+      existingBP.setEnabled(bId.isEnabled());
     }
-    bp.setEnabled(bId.isEnabled());
-    return bp;
   }
 
-  public Breakpoint addOrUpdate(final AsyncMessageReceiveBreakpoint bId) {
-    Breakpoint bp = truffleBreakpoints.get(bId);
+  public synchronized void addOrUpdate(final PromiseResolverBreakpoint bId) {
+    FullSourceCoordinate coord = bId.getCoordinate();
+    BreakpointEnabling<PromiseResolverBreakpoint> existingBP = promiseResolverBreakpoints.get(coord);
+    if (existingBP == null) {
+      existingBP = new BreakpointEnabling<PromiseResolverBreakpoint>(bId);
+      promiseResolverBreakpoints.put(coord, existingBP);
+    } else {
+      existingBP.setEnabled(bId.isEnabled());
+    }
+  }
 
+  public synchronized void addOrUpdate(final PromiseResolutionBreakpoint bId) {
+    FullSourceCoordinate coord = bId.getCoordinate();
+    BreakpointEnabling<PromiseResolutionBreakpoint> existingBP = promiseResolutionBreakpoints.get(coord);
+    if (existingBP == null) {
+      existingBP = new BreakpointEnabling<PromiseResolutionBreakpoint>(bId);
+      promiseResolutionBreakpoints.put(coord, existingBP);
+    } else {
+      existingBP.setEnabled(bId.isEnabled());
+    }
+  }
+
+  private void saveTruffleBasedBreakpoints(final SectionBreakpoint bId) {
+    Breakpoint bp = truffleBreakpoints.get(bId);
     if (bp == null) {
-      WebDebugger.log("RootBreakpoint: " + bId);
+      WebDebugger.log("SetSectionBreakpoint: " + bId);
       bp = Breakpoint.newBuilder(bId.getCoordinate().uri).
           lineIs(bId.getCoordinate().startLine).
           columnIs(bId.getCoordinate().startColumn).
           sectionLength(bId.getCoordinate().charLength).
           build();
-      bp.setCondition(BreakWhenActivatedByAsyncMessage.INSTANCE);
```

Comment:
@ctrlpz this is a problem, this line is essential :( and we really need tests for these things